src/parsers/gene_association_file.py

```diff
@@ -0,0 +1,117 @@
+    df["annotation_date"] = df["annotation_date"].astype(str)
+    
+    # Add constant metadata columns 
+    df["aggregator"] = "UniProt"
+    df["protocol_id"] = "NULL"
+    df["supporting_objects"] = df["supporting_objects"].fillna("NULL")
+    
+    # Define allowed predicates, determine negation, and clean predicate names
+    allowed_predicates = ["enables", "contributes_to", "acts_upstream_of_or_within", "involved_in",
+                          "acts_upstream_of", "acts_upstream_of_positive_effect",
+                          "acts_upstream_of_negative_effect", "acts_upstream_of_or_within_negative_effect",
+                          "acts_upstream_of_or_within_positive_effect", "located_in", "part_of", 
+                          "is_active_in", "colocalizes_with"]
+
+    df["negated"] = ~df["predicate"].isin(allowed_predicates)
+    df["predicate"] = df["predicate"].str.replace(r"^NOT\|", "", regex=True)
+    
+    # Load the external ECO mapping data
+    url = "https://raw.githubusercontent.com/evidenceontology/evidenceontology/master/gaf-eco-mapping.txt"
+    mapping_columns = ["Evidence_Code", "DB_Reference", "evidence_type"]
+    df2 = pd.read_csv(url, sep="\t", comment="#", header=None, names=mapping_columns)
+    
+    """
+    merge two daraframe
+    df1: go annotations data
+    df2: ECO mapping data
+    df2 DB_reference match with df1 publications
+    df2 evidence code match with df1 evidence code 
+    """
+
+    # Normalize strings for consistent matching 
+    df["publications"] = df["publications"].str.strip().str.upper()
+    df2["DB_Reference"] = df2["DB_Reference"].str.strip().str.upper()
+    df2["Evidence_Code"] = df2["Evidence_Code"].str.strip().str.upper()
+
+    # Merge input data with ECO mapping data
+    merged = df.merge(
+        df2,how="left",
+        left_on=["Evidence_Code", "publications"],
+        right_on=["Evidence_Code", "DB_Reference"])
+    
+    # Handle unmatched revidence_type by assigning default mappings
+    unmatched_mask = merged["evidence_type"].isna()
+    default_mapping = df2[df2["DB_Reference"].str.upper() == "DEFAULT"]
+
+    fallback_match = merged.loc[unmatched_mask, ["Evidence_Code"]].merge(default_mapping, how="left", on="Evidence_Code")
+    merged.loc[unmatched_mask, "evidence_type"] = fallback_match["evidence_type"].values
+
+    # Drop the unnecessary merged columns 
+    merged = merged.drop(columns=["DB_Reference"])
+    
+    # Save the processed annotations to a new CSV file
+    merged.to_csv(output_path, index=False)
+
+
+@click.command()
+@click.option('--input', '-i', required=True, help='Path to input GO annotation CSV file.')
+@click.option('--output', '-o', required=True, help='Path to output normalized CSV file.')
+
+
+def main(input, output):
+    """
+    CLI entry point to process GO annotations.
+
+    Args:
+        input (str): Input CSV file path.
+        output (str): Output CSV file path.
+    """
+
+    process_go_annotations(input, output)
```

Comment:
you might want to do some checks here before you start processing -- e.g. does the path to the output file exist? What about the path to the input file? How about the gaf-eco-mapping file?

I prefer to do all these checks at the start so that the script can bail out early if there's anything wrong, and I don't end up waiting for ages for it to run, only to find out that I misspelled a directory in the output file or some other stupid error!